extensions/functions_string.yaml

```diff
@@ -248,3 +248,192 @@ scalar_functions:
+          - value: "varchar<L2>"
+            name: "replacement"
+            description: The new string to insert at the start position.
+        return: "varchar<L3>"
+  -
+    name: lower
+    description: >-
+      Transform the string to lower case characters. Implementation should follow the utf8_unicode_ci
+      collations according to the Unicode Collation Algorithm described at http://www.unicode.org/reports/tr10/.
+    impls:
+      - args:
+          - value: "string"
+        return: "string"
+      - args:
+          - value: "varchar<L1>"
+        return: "varchar<L1>"
+      - args:
+          - value: "fixedchar<L1>"
+        return: "fixedchar<L1>"
+  -
+    name: upper
+    description: >-
+      Transform the string to upper case characters. Implementation should follow the utf8_unicode_ci
+      collations according to the Unicode Collation Algorithm described at http://www.unicode.org/reports/tr10/.
+    impls:
+      - args:
+          - value: "string"
+        return: "string"
+      - args:
+          - value: "varchar<L1>"
+        return: "varchar<L1>"
+      - args:
+          - value: "fixedchar<L1>"
+        return: "fixedchar<L1>"
+  -
+    name: swapcase
+    description: >-
+      Transform the string's lowercase characters to uppercase and uppercase characters to
+      lowercase. Implementation should follow the utf8_unicode_ci collations according to the
+      Unicode Collation Algorithm described at http://www.unicode.org/reports/tr10/.
+    impls:
+      - args:
+          - value: "string"
+        return: "string"
+      - args:
+          - value: "varchar<L1>"
+        return: "varchar<L1>"
+      - args:
+          - value: "fixedchar<L1>"
+        return: "fixedchar<L1>"
+  -
+    name: capitalize
+    description: >-
+      Capitalize the first character of the input string.
+    impls:
+      - args:
+          - value: "string"
+        return: "string"
+      - args:
+          - value: "varchar<L1>"
+        return: "varchar<L1>"
+      - args:
+          - value: "fixedchar<L1>"
+        return: "fixedchar<L1>"
+  -
+    name: title
+    description: >-
+      Converts the input string into titlecase. Capitalize the first character of each word in the
+      input string except for articles (a, an, the).
+    impls:
+      - args:
+          - value: "string"
+        return: "string"
+      - args:
+          - value: "varchar<L1>"
+        return: "varchar<L1>"
+      - args:
+          - value: "fixedchar<L1>"
+        return: "fixedchar<L1>"
+  -
+    name: length
+    description: Return the number of characters in the input string.
+    impls:
+      - args:
+          - value: "string"
+        return: i64
+      - args:
+          - value: "varchar<L1>"
+        return: i64
+      - args:
+          - value: "fixedchar<L1>"
+        return: i64
+  -
+    name: bit_length
+    description: Return the number of of bits in the input string.
```

Comment:
Yeah, it seems to be bits lol.  https://www.w3resource.com/PostgreSQL/bit-length-function.php

Reply:
The docs there don't list any reason why it's in bits rather than bytes... IMO it should just not exist in this form and should just be `byte_length`. When converting a `bit_length(x)` function in SQL to Substrait it can always be represented as `multiply(byte_length(x), 8)`, so `byte_length` is more primitive. I could see the reason they used bit granularity being that with different storage compression methods you might end up with bit-level stuff, but Substrait doesn't define any of that stuff for `string`s; all it says is that it's UTF-8-encoded, so at best you can have a function that specifies the number of bytes needed for said UTF-8 encoding.

OTOH, there's precedent for following SQL spec nonsense (like `sum([])` returning null). Is this specifically a Postgres function or is it part of the SQL spec?

Reply:
Postgres and MySQL have bit_length, they both also have one called 'octet_length', which returns bytes: 
https://www.w3resource.com/PostgreSQL/octet_length-function.php

I can add a separate byte_length function, so we have both.

Reply:
> The above function is a synonym for LENGTH().

Are their strings not UTF-8? I don't even know anymore.

I think I'm just going to abstain from this discussion because I increasingly feel like I don't have the background to make informed decisions about this. @jacques-n do you hold strong opinions about these functions?

Reply:
Yeah, it's confusing because i've seen length defined 2 different ways.  One returns the bytes, the other returns the number of characters (which is why I just updated the length function in this PR to be char_length).

Reply:
@jacques-n SQL Server and Oracle have both bit_length and octet_length (for bytes).  I've added octet_length to this PR as well.

https://docs.oracle.com/cd/B19188_01/doc/B15917/sqfunc.htm#i1005968
https://docs.microsoft.com/en-us/sql/odbc/reference/appendixes/string-functions?view=sql-server-ver16